src/core/lib/iomgr/ev_epoll_linux.c

```diff
@@ -1060,101 +1095,182 @@ static void fd_orphan(grpc_exec_ctx *exec_ctx, grpc_fd *fd,
   GRPC_ERROR_UNREF(error);
 }
 
-static grpc_error *fd_shutdown_error(grpc_fd *fd) {
-  if (!fd->shutdown) {
-    return GRPC_ERROR_NONE;
-  } else {
-    return GRPC_ERROR_CREATE_REFERENCING("FD shutdown", &fd->shutdown_error, 1);
+static void notify_on(grpc_exec_ctx *exec_ctx, grpc_fd *fd, gpr_atm *state,
+                      grpc_closure *closure) {
+  bool is_done = false;
+  while (!is_done) {
```

Comment:
ok - what you wrote does look a bit cleaner. I'll do it.  FWIW, I am very scared of having `return` statements in between the functions and always strive to have just one `return` at the end (and sometimes an extra `return`  at the beginning after some initial-condition validations)